app/background/wallet/service.js

```diff
@@ -152,9 +149,123 @@ class WalletService {
     }
   };
 
+  getAddresses = async (depth = 10000) => {
+    await this._ensureClient();
+    const wdb = this.node.wdb;
+    const wallet = await wdb.get(WALLET_ID);
+    const account = await wallet.getAccount('default');
+
+    const addresses = Array(depth)
+      .fill(0)
+      .map((_, i) => {
+        const receive = account.deriveReceive(i).getAddress().toString();
+        const change = account.deriveChange(i).getAddress().toString();
+        return [receive, change];
+      })
+      .reduce((acc, [receive, change]) => {
+        acc.push(receive);
+        acc.push(change);
+        return acc;
+      }, []);
+
+    return addresses;
+  };
+
+  getTXByAddresses = async () => {
+    await this._ensureClient();
+    const addresses = await this.getAddresses();
+    let txs = [];
+
+    for (let i = 0; i < 20000; i = i + 1000) {
+      const transactions = await nodeService.getTXByAddresses(addresses.slice(i, i+1000));
+      txs = txs.concat(transactions);
+    }
+
+    return txs;
+  };
+
   rescan = async (height = 0) => {
     await this._ensureClient();
     const wdb = this.node.wdb;
+
+    dispatchToMainWindow({type: START_SYNC_WALLET});
+    dispatchToMainWindow({
+      type: SYNC_WALLET_PROGRESS,
+      payload: parseInt(0),
+    });
+
+    let transactions = await this.getTXByAddresses();
+    transactions = transactions.sort((a, b) => {
+      if (a.index > b.index) return 1;
+      if (b.index > a.index) return -1;
```

Comment:
I think `index` of a TXMeta is its position in a block (for example coinbase index is `0`, etc) so I'm not sure what the advantage is to sorting here.

Reply:
It is basically pre-sorting by transactions index, before indexing them by block height. Indexing them by block height first AND THEN sort do seem more efficient though -- I will make that change.